File: 2024/March/3-17 57 Insert Interval/InsertInterval.py
from typing import List

class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        if not intervals:
            return [newInterval]
        
        starts = []
        ends = []
        solution = []

        for interval in intervals:
            starts.append(interval[0])
            ends.append(interval[1])

        starts.append(newInterval[0])
        ends.append(newInterval[1])

        starts.sort()
        ends.sort()
        
        while starts and ends:
            start = starts.pop(0)
            end = ends.pop(0)
            while starts and ends and end >= starts[0]:
                starts.pop(0)
                end = ends.pop(0)
            solution.append([start, end])

        return solution

File: 2024/March/3-17 57 Insert Interval/InsertInterval.js
/**
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */
var insert = function(intervals, newInterval) {
    let starts = [];
    let ends = [];
    let solution = [];

    for(let interval of intervals) {
        starts.push(interval[0]);
        ends.push(interval[1]);
    }

    starts.push(newInterval[0]);
    ends.push(newInterval[1]);

    starts.sort((a,b) => (a - b));
    ends.sort((a,b) => (a - b));

    console.log(starts);
    console.log(ends);

    while(starts.length > 0) {
        let start = starts.shift();
        let end = ends.shift();
        while(starts.length > 0 && end >= starts[0]) {
            starts.shift();
            end = ends.shift();
        }
        solution.push([start, end]);
    }

    return solution;
};